File: apps/cli/src/commands/my.ts
import { Command } from "commander";
import { getClient, getTenantId } from "../api/client";
import { getOutputOptions } from "../utils/command";
import type { OutputOptions } from "../utils/output";
import { print } from "../utils/output";
import { unwrapApiResult } from "../utils/errors";
import { parseTaskRef, resolveProject } from "../utils/projects";
import { findDoneStatusId } from "../utils/statuses";

type MyCommandOptions = OutputOptions & {
  filter?: string;
};

export function registerMyCommands(program: Command): void {
  const my = program.command("my").description("My Tasks commands");

  my
    .command("list")
    .description("List tasks assigned to me")
    .option(
      "--filter <filter>",
      "today | week | no_due_date | overdue | all",
      "all",
    )
    .action(async (opts: MyCommandOptions, cmd) => {
      const output = getOutputOptions(cmd);
      const client = getClient();
      const tenantId = getTenantId();
      const result = await client.GET("/v1/tenants/{tenant_id}/users/me/tasks", {
        params: {
          path: { tenant_id: tenantId },
          query: { filter: opts.filter ?? "all" },
        },
      });
      print(unwrapApiResult(result), output);
    });

  my
    .command("complete")
    .description("Complete a personal or assigned task by ref (e.g. ME-3)")
    .argument("<ref>", "Task ref (ME-N, KEY-N, or UUID)")
    .action(async (ref: string, _opts, cmd) => {
      const output = getOutputOptions(cmd);
      const parsed = parseTaskRef(ref);
      const client = getClient();
      const tenantId = getTenantId();

      if ("uuid" in parsed) {
        let matchedTask: import("../api/paths").MyTaskItem | undefined;
        const PAGE = 200;
        let offset = 0;
        while (!matchedTask) {
          const list = await client.GET("/v1/tenants/{tenant_id}/users/me/tasks", {
            params: {
              path: { tenant_id: tenantId },
              query: { filter: "all", limit: PAGE, offset },
            },
          });
          const tasks = unwrapApiResult(list).tasks;
          matchedTask = tasks.find((t) => t.id === parsed.uuid);
          if (matchedTask || tasks.length < PAGE) break;
          offset += PAGE;
        }
        if (!matchedTask) {
          const { handleApiError } = await import("../utils/errors");
          handleApiError({ status: 404, message: `Task not found: ${ref}` });
        }
        const statusId = await findDoneStatusId(matchedTask!.project.id);
        const result = await client.PUT(
          "/v1/tenants/{tenant_id}/projects/{project_id}/tasks/{id}",
          {
            params: {
              path: {
                tenant_id: tenantId,
                project_id: matchedTask!.project.id,
                id: matchedTask!.id,
              },
            },
            body: { status_id: statusId },
          },
        );
        print(unwrapApiResult(result), output);
        return;
      }

      const project = await resolveProject(parsed.projectKey);
      const statusId = await findDoneStatusId(project.id);
      const result = await client.PUT(
        "/v1/tenants/{tenant_id}/projects/{project_id}/tasks/{id}",
        {
          params: {
            path: {
              tenant_id: tenantId,
              project_id: project.id,
              id: parsed.taskId,
            },
          },
          body: { status_id: statusId },
        },
      );
      print(unwrapApiResult(result), output);
    });
}

File: apps/backend/crates/cli/src/commands/my.rs
//! 自分のタスク。

use payload::my_tasks::{MyTaskItem, MyTasksListResponse};
use payload::tasks::TaskDetailResponse;

use crate::Context;
use crate::api::ApiClient;
use crate::cli::MyCommand;
use crate::error::{CliError, Result};
use crate::output::{OutputOptions, print};
use crate::resolve::{TaskRef, find_done_status_id, parse_task_ref, resolve_project};

/// UUID から自分のタスクを探すときの 1 回あたりの取得数。
const PAGE: u64 = 200;

pub async fn run(context: &Context, command: MyCommand, output: OutputOptions) -> Result<i32> {
    match command {
        MyCommand::List { filter } => {
            let api = &context.connect()?;
            let tasks: MyTasksListResponse =
                api.get(&my_tasks_path(api), &[("filter", filter)]).await?;
            print(&tasks, output);
        }
        MyCommand::Complete { task_ref } => {
            // 参照の形は送る前に確かめる（設定不足より先に、直せる誤りを出す）
            let parsed = parse_task_ref(&task_ref)?;
            let api = &context.connect()?;
            let (project_id, task_id) = match parsed {
                TaskRef::Uuid(uuid) => {
                    let task = find_my_task(api, &uuid.to_string()).await?;
                    (task.project.id, task.id.to_string())
                }
                TaskRef::Seq {
                    project_key,
                    task_id,
                } => (resolve_project(api, &project_key).await?.id, task_id),
            };
            let status_id = find_done_status_id(api, project_id).await?;
            let project_id = project_id.to_string();
            let updated: TaskDetailResponse = api
                .put(
                    &[
                        "v1",
                        "tenants",
                        api.tenant_id(),
                        "projects",
                        &project_id,
                        "tasks",
                        &task_id,
                    ],
                    &crate::commands::tasks::done_request(status_id),
                )
                .await?;
            print(&updated, output);
        }
    }
    Ok(0)
}

fn my_tasks_path(api: &ApiClient) -> [&str; 6] {
    ["v1", "tenants", api.tenant_id(), "users", "me", "tasks"]
}

/// UUID だけでは所属プロジェクトが分からないので、自分のタスクを辿って突き合わせる。
async fn find_my_task(api: &ApiClient, id: &str) -> Result<MyTaskItem> {
    let mut offset = 0u64;
    loop {
        let page: MyTasksListResponse = api
            .get(
                &my_tasks_path(api),
                &[
                    ("filter", "all".into()),
                    ("limit", PAGE.to_string()),
                    ("offset", offset.to_string()),
                ],
            )
            .await?;
        let count = page.tasks.len() as u64;
        if let Some(task) = page
            .tasks
            .into_iter()
            .find(|task| task.id.to_string() == id)
        {
            return Ok(task);
        }
        if count < PAGE {
            return Err(CliError::not_found(format!("Task not found: {id}")));
        }
        offset += PAGE;
    }
}
